Week 3 libraries tris

dacayant opened the pull request:
Sorry for the late submission. I had a lot of family troubles this week.

imichaelnorris reviewed (commented):
When I upload an image on http://tdacayan-sps-summer21.appspot.com/ it fails with a 403 error which I pasted below. I think this can be fixed by adding the storage.objects.create to your account in the Google Cloud console

Did you test this with a local running cloud storage instance?

com.google.cloud.storage.StorageException: 403 Forbidden POST https://storage.googleapis.com/upload/storage/v1/b/tdacayan-sps-summer21.appspot.com/o?projection=full&uploadType=multipart { "error": { "code": 403, "message": "tdacayan-sps-summer21@appspot.gserviceaccount.com does not have storage.objects.create access to tdacayan-sps-summer21.appspot.com/windows_stock.jpeg.", "errors": [ { "message": "tdacayan-sps-summer21@appspot.gserviceaccount.com does not have storage.objects.create access to tdacayan-sps-summer21.appspot.com/windows_stock.jpeg.", "domain": "global", "reason": "forbidden" } ] } }

dacayant commented:
Strangely enough, it was working when I tested it on the dev server. I'll try add the storage.objects.create.

imichaelnorris commented:
> Strangely enough, it was working when I tested it on the dev server. I'll try add the storage.objects.create.

Hmm that's strange! I downloaded the example image from your portfolio and uploaded it. Is it possible that you had already uploaded an image with the same name? And then maybe the error that I saw was actually it trying to overwrite the existing image and failing.

dacayant commented:
I don't think so because I am also getting the same error now after trying to upload a new image. Also, where/how do I add the storage.objects.create permission?

imichaelnorris commented:
First, did you enable the API here? https://console.cloud.google.com/apis/api/datastore.googleapis.com/overview

If it still doesn't work after that, I *think* the permission is added in the iam page: https://console.cloud.google.com/iam-admin/iam but going to the correct service account, clicking the "edit icon" in the inheritance column, and then click "add role", and then search for "Cloud Datastore", and add the "Cloud Datastore Owner" role.

dacayant commented:
I can't find the Cloud Datastore role for my permissions. I think it might be because I'm using Cloud Storage and not DataStore.
![image](https://user-images.githubusercontent.com/84953909/122839271-b8750880-d2c5-11eb-8810-6c4b6ea9496c.png)

dacayant commented:
I fixed the permissions. I just added the Cloud Storage - Storage Admin permission to allUsers in the cloud storage browser. The file submission should work now.

imichaelnorris commented on portfolio/src/main/java/com/google/sps/servlets/ImageFormHandler.java:
```diff
@@ -0,0 +1,64 @@
+    }
+    /** Uploads a file to Cloud Storage and returns the uploaded file's URL. */
+    private static String uploadToCloudStorage(String fileName, InputStream fileInputStream) {
+        String projectId = "tdacayan-sps-summer21";
```
projectId, bucketName, and storage can all be made static variables of this class.

imichaelnorris reviewed (approved):
Glad you were able to fix it on your own. The pull request looks good to me.

I left one small comment about moving variables to be static variables.

Also sorry about suggesting the wrong API, I get confused about all of the Cloud APIs with different names (and I was using the one I suggested in a personal project so it was on my mind).

dacayant commented:
How do I resolve this conflict? I've tried to download Git Bash for the command line but I wasn't able to merge this branch into the main branch.

imichaelnorris commented:
It looks like there's a merge conflict between buttons.js in main and in this branch.

What git command did you run when you used Git Bash?

What is the output from "git status" ?

dacayant commented:
I tried doing these commands but nothing really happened.
![image](https://user-images.githubusercontent.com/84953909/122980744-042bbe80-d367-11eb-9e03-5ea745113ef1.png)

As for git status, it says that my one branch is 13 commits ahead of my main branch.

dacayant commented:
When i get to the last step, I get this error message:
![image](https://user-images.githubusercontent.com/84953909/122981871-38ec4580-d368-11eb-93eb-65becf013f1b.png)

imichaelnorris commented:
Oh, that's strange. I thought it was an issue with something else entirely.

It says that an approving review is needed but I approved the changes, so I'm pretty confused.

I'm not sure if this is the cause, but I did leave a comment which is unresolved, so I'll resolve that comment to see if it is the cause and then you can try again.

dacayant commented:
I tried it again and it still didn't work. 

I was thinking, if I were to just change the file name directly to match the new file rename, would that solve the conflict?

imichaelnorris commented:
I think the issue has to do with the rename of buttons.js to interactive-elements.js, so if you renamed interactive-elements.js back to buttons.js in this pull request then I think this issue would go away.

It should be possible to rename a file in one pull request and not have a merge conflict when merging it back in, I would expect that to be what happens if you run "git mv buttons.js interactive-elements.js"

dacayant commented:
Oh. I didn't know that was a command in git. I had been using the commit making tab in Google Shell for the past couple weeks so that must have been why there was a conflict. I'll remember that the next time I try to rename a file.
